beacon-chain/db/kv/attestations.go

```diff
@@ -1,44 +1,173 @@
-// TODO(#3164): Implement.
 func (k *Store) Attestation(ctx context.Context, attRoot [32]byte) (*ethpb.Attestation, error) {
-	return nil, nil
+	att := &ethpb.Attestation{}
+	err := k.db.View(func(tx *bolt.Tx) error {
+		bkt := tx.Bucket(attestationsBucket)
+		c := bkt.Cursor()
+		for k, v := c.Seek(attRoot[:]); k != nil && bytes.Contains(k, attRoot[:]); k, v = c.Next() {
+			if v != nil {
+				return proto.Unmarshal(v, att)
+			}
+		}
+		return nil
+	})
+	return att, err
 }
 
 // Attestations retrieves a list of attestations by filter criteria.
-// TODO(#3164): Implement.
 func (k *Store) Attestations(ctx context.Context, f *filters.QueryFilter) ([]*ethpb.Attestation, error) {
-	return nil, nil
+	atts := make([]*ethpb.Attestation, 0)
+	hasFilterSpecified := !reflect.DeepEqual(f, &filters.QueryFilter{}) && f != nil
+	err := k.db.View(func(tx *bolt.Tx) error {
+		bkt := tx.Bucket(attestationsBucket)
+		c := bkt.Cursor()
+		for k, v := c.First(); k != nil; k, v = c.Next() {
+			if v != nil && (!hasFilterSpecified || attestationFilterCriteria(k, f)) {
+				att := &ethpb.Attestation{}
+				if err := proto.Unmarshal(v, att); err != nil {
+					return err
+				}
+				atts = append(atts, att)
+			}
+		}
+		return nil
+	})
+	return atts, err
 }
 
 // HasAttestation checks if an attestation by root exists in the db.
-// TODO(#3164): Implement.
 func (k *Store) HasAttestation(ctx context.Context, attRoot [32]byte) bool {
-	return false
+	exists := false
+	// #nosec G104. Always returns nil.
+	k.db.View(func(tx *bolt.Tx) error {
+		bkt := tx.Bucket(attestationsBucket)
+		c := bkt.Cursor()
+		for k, v := c.Seek(attRoot[:]); k != nil && bytes.Contains(k, attRoot[:]); k, v = c.Next() {
+			if v != nil {
+				exists = true
+				return nil
+			}
+		}
+		return nil
+	})
+	return exists
 }
 
 // DeleteAttestation by root.
-// TODO(#3164): Implement.
 func (k *Store) DeleteAttestation(ctx context.Context, attRoot [32]byte) error {
-	return nil
+	return k.db.Update(func(tx *bolt.Tx) error {
+		bkt := tx.Bucket(attestationsBucket)
+		c := bkt.Cursor()
+		for k, v := c.Seek(attRoot[:]); k != nil && bytes.Contains(k, attRoot[:]); k, v = c.Next() {
+			if v != nil {
+				return bkt.Delete(k)
+			}
+		}
+		return nil
+	})
 }
 
 // SaveAttestation to the db.
-// TODO(#3164): Implement.
 func (k *Store) SaveAttestation(ctx context.Context, att *ethpb.Attestation) error {
-	return nil
+	key, err := generateAttestationKey(att)
+	if err != nil {
+		return err
+	}
+	enc, err := proto.Marshal(att)
+	if err != nil {
+		return err
+	}
+	return k.db.Update(func(tx *bolt.Tx) error {
+		bucket := tx.Bucket(attestationsBucket)
+		return bucket.Put(key, enc)
+	})
 }
 
 // SaveAttestations via batch updates to the db.
-// TODO(#3164): Implement.
 func (k *Store) SaveAttestations(ctx context.Context, atts []*ethpb.Attestation) error {
-	return nil
+	return k.db.Batch(func(tx *bolt.Tx) error {
+		bucket := tx.Bucket(attestationsBucket)
+		for _, att := range atts {
```

Comment:
can you generate the keys outside of the boltdb callback ? if not there is a lock on the db while you are looping through all the attestations